src/context.rs

```diff
@@ -1237,52 +1266,229 @@ impl SeafowlContext for DefaultSeafowlContext {
+                                            );
+                                    }
+                                    Err(error) => warn!(
+                                        "Failed constructing pruning statistics for table {} (version: {}) during UPDATE execution: {}",
+                                        table.name, table.table_version_id, error
+                                    )
+                                }
+                            }
+
+                            let schema = table.schema().as_ref().clone();
+                            let df_schema =
+                                table.schema.arrow_schema.clone().to_dfschema()?;
+
+                            // Create a complete projection expression, including replicating
+                            // the columns omitted in the provided assignments
+                            let projection_exprs = schema
+                                .fields
+                                .iter()
+                                .enumerate()
+                                .map(|(ind, f)| {
+                                    if assignments.contains_key(f.name()) {
+                                        let mut expr = create_physical_expr(
+                                            &assignments[f.name()],
+                                            &df_schema,
+                                            &schema,
+                                            &ExecutionProps::new(),
+                                        )?;
+                                        let data_type = f.data_type().clone();
+                                        if expr.data_type(&schema)? != data_type {
+                                            // Literal value potentially mistyped; try to re-cast it
+                                            expr = try_cast(expr, &schema, data_type)?;
+                                        }
+                                        Ok((expr, f.name().to_string()))
+                                    } else {
+                                        // Just replicate the omitted columns
+                                        Ok((
+                                            Arc::new(ColumnExpr::new(f.name(), ind))
+                                                as Arc<dyn PhysicalExpr>,
+                                            f.name().to_string(),
+                                        ))
+                                    }
+                                })
+                                .collect::<Result<Vec<(Arc<dyn PhysicalExpr>, String)>>>(
+                                )?;
+
+                            let mut final_partition_ids =
+                                Vec::with_capacity(partitions.len());
+
+                            // Group adjacent partitions eligible for updating
+                            let partitions_grouped: Vec<(bool, Vec<SeafowlPartition>)> =
```

Comment:
In principle, if the ordering is not important, then we can (here as well as for DELETE) divide all the partitions into exactly two groups, and re-process the group that needs to be changed in one go.

Reply:
Kinda true (especially since we're reordering data inside of the partitions anyway, since we're merging changed with the unchanged rows), but we have this code already, so I don't mind it staying here.